tldraw/tldraw
# Add an onCrop handler to ShapeUtil

This is called when cropping and allows you to override the crop calculations. This is useful for instance to set a min width and/or height for the cropping.

It can also be used to provide a completely different cropping mechanism, e.g. to keep the size of the shape fixed and instead resize the image when cropping. This would require overriding the SelectionForeground component, but it's better to only have to override that than the whole selection tool. And maybe that component can be made more dynamic in the future.

This also defines specific crop interfaces instead of using the ones from the image shape and uses the correct util instead of the image util when cropping. Any shape can define that it supports cropping, so I don't think it should be tied to the image shape.

### Change type


### Test plan

1. Add an onCrop handler to a util for a shape that can be cropped.
1. Create a shape with a crop prop.
2. Try to crop the shape and see that the behavior in onCrop is applied.


### Release notes

- Add support for an onCrop handler on shape utils that allows you to prevent or modify the crop.
- The `TLImageShapeCrop` type has been replaced by `TLShapeCrop`.
